### ISSUE TITLE
require_tenant, but allow explicit disabling

### ISSUE THREAD

**Issue by lowjoel** [ErwinM/acts_as_tenant#125]
I have a need for require_tenant to be true (to prevent leakages), but also at the same time I have a global admin panel where I do not want to have tenanting for a global overview.

Currently, the default_tenant is set, then unset to nil after every request. Would it be a good idea to explicitly support `false` in addition to a valid instance, to specify explicitly that I have explicitly set a tenant, just that it's all of them?

Kind of like how cancan(can) lets you specify a require authorisation flag, but lets you silence that when you know what you're doing

It is possible to unset the require tenant flag, but I think that's changing a global configuration for local effect which is... eww.

**Comment by duboff** [comment]
What we did is disable require_tenant in our admin controller:
```rb
before_action :disable_tenant

def disable_tenant
   ActsAsTenant.configuration.require_tenant = false
end
```
It doesn't affect other controllers. I guess it's not very clean but it works.

**Comment by lowjoel** [comment]
Unless my understanding of the code is wrong, that would definitely affect other controllers:

`ActsAsTenant.configuration` references `@@configuration`, a class variable, that's only instantiated once. So unless you've got separate forks of the application and one only handles admin requests, if the same process subsequently handles a request from another part of the app, require_tenant will still be false.

Did I miss something?

**Comment by duboff** [comment]
I think it might affect other controllers if the admin himself then goes to other routes from the admin panel. But I just checked and for other users or even for admin coming from a new request, the global setting is reset. In our app this does not happen as the rest of the app is in a different app consuming only an API from the main one.

**Comment by duboff** [comment]
On second thought you are right.

**Comment by gommo** [comment]
I'm after something similar. Anyone come up with an easy way to do this. I'd love our site-admin account to be able to view all models, preferably readonly.

**Comment by duboff** [comment]
I made an attempt in #131, would love to hear what you think. A simpler solution when it is available is to just do Model.unscoped in the admin controllers.

**Comment by lleolin** [comment]
Brotip™: use `unscope(where: :<tenant_id>)` instead of `unscoped`. The latter obliterates any and all extant scoping prior to that call, which can be problematic for chaining. The former just eliminates the tenant scoping.

**Comment by duboff** [comment]
Thanks @lleolin, that indeed is better.

**Comment by duboff** [comment]
@lleolin If you are referring to #131, then it's not required since we just avoid adding scope in the first place, rather than unset it. Unscoped is just a property on the ActsAsTenant, that doesn't do anything.

But your method might be a good enough solution to the problem so we don't need a special API.

**Comment by lleolin** [comment]
@duboff I just meant as an alternative to `Model.unscoped`. It makes sense to avoid the scoping than to remove it, but I had to work with the existing implementation of the official gem for now.

That said, even with `unscope(where: :<tenant_id>)`, you can potentially eliminate part of a Relation that you might have wanted, so the order of operations (so to speak) still matters, but it's just a little more precise.

**Comment by lowjoel** [comment]
I believe there were bugs in Rails where default scopes didn't get properly unscoped when using `unscope` or even `unscoped` (can't remember the issue numbers.) I am not entirely sure if they have been fixed yet, but I do know that they were incredibly annoying to work around.

**Comment by ErwinM** [comment]
closed by #131

**Comment by lowjoel** [comment]
Awesome, thanks @duboff and @ErwinM
